NFT/src/main.sw

```diff
@@ -0,0 +1,368 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `from` - The `Identity` which currently owns the token to be transfered.
+    /// * `to` - The `Identity` which the ownership of the token should be set to.
+    /// * `token_id` - The `u64` id of the token which should be transfered.
+    ///
+    /// # Reverts
+    ///
+    /// * When the `token_id` does not map to an existing token.
+    /// * When the sender is not the `owner`.
+    /// * When the sender is not approved to transfer the `token_id` on the owner's behalf.
+    /// * When the sender is not approved to transfer all tokens on the owner's behalf.
+    #[storage(read, write)]fn transfer_from(from: Identity, to: Identity, token_id: u64) {
+        // Make sure the `token_id` maps to an existing token
+        let meta_data = storage.meta_data.get(token_id);
+        require(meta_data.is_some(), InputError::TokenDoesNotExist);
+
+        // Ensure that the sender is either:
+        // 1. The owner of the token
+        // 2. Approved for transfer of this `token_id`
+        // 3. Or an operator and the token is owned by the owner
+        let sender = msg_sender().unwrap();
+        let mut meta_data = meta_data.unwrap();
+        let approved = meta_data.approved;
+        require(sender == meta_data.owner || (approved.is_some() && sender == approved.unwrap()) || (from == meta_data.owner && storage.operator_approval.get(sha256((from, sender)))), AccessError::SenderNotOwnerOrApproved);
+
+        // Set the new owner of the token and reset the approved Identity
+        meta_data.owner = to;
+        meta_data.approved = Option::None();
+        storage.meta_data.insert(token_id, Option::Some(meta_data));
+
+        // Decrease the previous owner's balance of tokens
+        storage.balances.insert(from, storage.balances.get(from) - 1);
+
+        // Increase the new owner's balance of tokens
+        storage.balances.insert(to, storage.balances.get(to) + 1);
+
+        // Log the transfer event
+        log(TransferEvent {
+            from, sender, to, token_id
+        });
+    }
+
+    /// Gives approval to the 'approved' `Identity` to transfer the specified token on the owner's behalf.
+    ///
+    /// # Arguments
+    ///
+    /// * `approved` - The `Identity` which will be allowed to transfer the token.
+    /// * `token_id` - The specific token which the owner is giving approval to.
+    /// * `approve` - The `bool` which wil allow or disallow transfers.
+    ///
+    /// # Reverts
+    ///
+    /// * When `token_id` does not map to an existing token.
+    /// * When the 'approved' `Identity` is the token's owner.
+    /// * When the sender is not the token's owner.
+    #[storage(read, write)]fn approve(approved: Option<Identity>, token_id: u64) {
+        // Ensure this is a valid token
+        let meta_data = storage.meta_data.get(token_id);
+        require(meta_data.is_some(), InputError::TokenDoesNotExist);
```

Comment:
Utility functions don't inherently have access to storage. I could add something like this however:

```
pub fn token_metadata(meta_data: Option<MetaData>) -> MetaData{
    require(meta_data.is_some(), InputError::TokenDoesNotExist);
    meta_data.unwrap()
}
```

Where calling it would look like `let mut meta_data = token_metadata(storage.meta_data.get(token_id));`

Reply:
I personally think this would be preferable, but feel free to leave it as it is.

Reply:
Just tried it out and it lowers the contract size by a few bytes so I'll take it!